Add support for EPICS Archiver Appliance

rneswold opened the pull request:
This PR adds support to the `acceleratorData` subscription to retrieve historical PV data from the Archiver Appliance.

rneswold committed:
:arrow_up: add necessary crates and get latest crates

rneswold committed:
:arrow_up: add necessary crates and get latest crates

rneswold committed:
:sparkles: initial EPICS archive appliance support

rneswold committed:
:sparkles: initial EPICS archive appliance support

rneswold committed:
Merge branch 'main' into epics

rneswold committed:
:bug: used the wrong name for field

rneswold committed:
:arrow_up: use latest protocol definitions

rneswold committed:
:recycle: split custom streams into separate modules

rneswold committed:
:sparkles: add a stream that accumulates scalar readings

The EPICS archiver sends one point at a time which is too slow for the
plotting app (the plotting app does a lot of work each time it gets a
packet of data.) This stream combinator will be used to buffer up a
large number of points to amortize the cost of updating the graph.

rneswold committed:
:art: make MAX_PAYLOAD a const generic

Since we're only going to ever instantiate one instance of this stream
type, there won't be code bloat by using generic parameters. However, I
can instantiate a max payload of 4, for instance, when writing unit
tests.

rneswold committed:
Merge remote-tracking branch 'origin/epics' into epics

rneswold committed:
:art: use clippy recommendation

rneswold committed:
:zap: improve stream code

Due to inexperience, I was boxing each stream that was built. Each
combinator I created added another heap allocation and required a
dereference when pulling elements.

With the help of an AI, the stream APIs have been refactored to use
generics and `impl Stream<>` monomorphisms to create combined streams
that don't require heap allocations (besides any underlying vectors they
might use in their state.)

rneswold committed:
:sparkles: convert `GroupScalars` to non-boxed version

rneswold committed:
:white_check_mark: adds unit tests for `GroupScalars`

beauremus commented on src/graphql/acsys/mod.rs:
```diff
@@ -417,26 +486,59 @@ impl<'ctx> ACSysSubscriptions {
+        device: &str, start_time: f64, end_time: f64,
+    ) -> Result<impl Stream<Item = global::DataReply> + Send + 'static + Unpin>
+    {
+        const BASE_URL: &str =
```
Is this temporary?
We want config like this defined in the K8s config and not code. Via a envar.
Also, no SSL.
Do both these get resolved when we move to gRPC?

beauremus commented on src/graphql/acsys/datastream/datamerge.rs:
```diff
@@ -0,0 +1,307 @@
+    task::{Context, Poll},
+};
+use tracing::warn;
+
```
This is tough to grok.
I'm a bit concerned about absurd user historical requests bringing a pod to its knees due to buffering.

I think we could leverage concurrency and separation of concerns a bit more to make this simpler to understand. (concurrency == simpler 🤣)

Here's what I worked with copilot to mock up:
1. Stream processing:
```
use tokio::sync::mpsc;
use tokio::task;

async fn process_archived_data(
    archived: impl Stream<Item = global::DataReply> + Send + 'static + Unpin,
    tx: mpsc::Sender<(i32, global::DataReply)>,
) {
    tokio::pin!(archived);
    while let Some(reply) = archived.next().await {
        if tx.send((reply.ref_id, reply)).await.is_err() {
            break; // Receiver dropped
        }
    }
}

async fn process_live_data(
    live: impl Stream<Item = global::DataReply> + Send + 'static + Unpin,
    tx: mpsc::Sender<(i32, global::DataReply)>,
) {
    tokio::pin!(live);
    while let Some(reply) = live.next().await {
        if tx.send((reply.ref_id, reply)).await.is_err() {
            break; // Receiver dropped
        }
    }
}
```
2. Merging
```
async fn merge_streams(
    rx: mpsc::Receiver<(i32, global::DataReply)>,
) -> impl Stream<Item = global::DataReply> {
    tokio_stream::wrappers::ReceiverStream::new(rx)
        .filter_map(|(ref_id, reply)| async move {
            // Apply any filtering or ordering logic here
            Some(reply)
        })
}
```
3. Execution
```
async fn data_merge(
    archived: impl Stream<Item = global::DataReply> + Send + 'static + Unpin,
    live: impl Stream<Item = global::DataReply> + Send + 'static + Unpin,
) -> impl Stream<Item = global::DataReply> {
    let (tx, rx) = mpsc::channel(100);

    // Spawn tasks for processing streams
    task::spawn(process_archived_data(archived, tx.clone()));
    task::spawn(process_live_data(live, tx));

    // Merge the streams
    merge_streams(rx).await
}
```

You've clearly thought about this a lot more than me. So don't let me hold this up, but I hope this plants some seeds.

beauremus reviewed (commented):
I agree with @jacob-curley-fnal's comments.
I'm willing to approve, but this is so much that I'd like more eyes on it and ideally some high-level discussion about structures.
Let me know how you want to proceed @rneswold.

rneswold commented:
> I'm having good luck returning an instance of `tokio_stream::wrappers::BroadcastStream` for alarms data.

The gRPC API returns a `Stream<>` and the GraphQL crate wants a `Stream<>`. With these combinators, my GraphQL resolvers build a `Stream<>` that processes and transforms the data and hands the stream off to the webserver. When the subscription is cancelled, the `Stream<>` is dropped and all resources are automatically cleaned up.

Starting a background task to put stream entries into a `mpsc::channel` and then have another read the channel and forward to the GraphQL stream doesn't seem as simple and elegant as using these stream combinators. When a client cancels a subscription, how do you reliably clean-up the async tasks? 

However, making the code easier to understand is a worthy goal.

I appreciate everyone's thoughts and comments!

rneswold commented:
> I agree with @jacob-curley-fnal's comments.
> I'm willing to approve, but this is so much that I'd like more eyes on it and ideally some high-level discussion about structures.
> Let me know how you want to proceed @rneswold.

We could Zoom or meet in the Cooler on a day that Beau and I are onsite to go through it in more detail.

rneswold committed:
:arrow_up: use latest protocols

rneswold committed:
Merge branch 'main' into epics

rneswold committed:
Merge branch 'main' into epics

rneswold committed:
:arrow_up: re-add `async-stream` to `Cargo.toml`

rneswold committed:
:white_check_mark: add tests for `EndOnDate`

This should make it reach 100% coverage.

rneswold committed:
:white_check_mark: add tests to `FilterDupes`

rneswold committed:
:white_check_mark: add one more test to `FilterDupes`

This should make the coverage 100%.

rneswold committed:
:art: format code

rneswold committed:
:wrench: add a timeout to gRPC Clock service

This adds a 2 second timeout for connecting to the clock service. Before
this, if the clock service wasn't running, it would take a long time
before the request would time-out.

rneswold committed:
:construction: add needed crates

rneswold committed:
:zap: improve return speed for EPICS archived data

The EPICS archiver appliance uses HTTP 1.1 to return the data.
Initially, the GraphQL resolver had to wait for the whole document to be
delivered before it could parsed and sent on to the client.

The appliance returns the document in chunked format. With the help of
AI, this commit is able to do a little hackery with the incoming stream
of chunks. It is able to parse the data returns as they arrive which
greatly speeds up the plotting of the data.

rneswold committed:
:white_check_mark: fix unit tests

The error messages changed in the `tonic` crate. This commit fixes the
tests to expect the new messages. I also removed a test that no longer
failed.

rneswold committed:
:art: use clippy recommendations

rneswold committed:
:arrow_up: use latest protocols

rneswold committed:
:white_check_mark: restore removed test

jacob-curley-fnal commented:
Look like a linting error is still hanging around:
```
struct `DeviceList` is never constructed
  --> src/g_rpc/generated/services.daq.rs:58:12
   |
58 | pub struct DeviceList {
   |            ^^^^^^^^^^
```

rneswold committed:
:wrench: remove chatty messages

rneswold committed:
:art: add semicolon

When I restored the test, I didn't add the semicolon like the original code.

rneswold committed:
Merge branch 'main' into epics

rneswold committed:
:arrow_up: use latest, fixed protocols

beauremus commented:
> Right now I consider this project to be in a broken state. There are a lot of great contributions being made, but they are breaking our "operational" GraphQL service.
> 
> @jacob-curley-fnal: merging with `main` pulled in your `.cargo/config.toml` changes full of illegal URLs so I wasn't able to easily run the GraphQL service. I think these are good additions, but you're forgetting that we don't yet have a deployed, operational service -- it's still being manually run in my home directory as a development service.
> 
> We need to get the Kubernetes deployment working sooner than later so we all can be contributing using the same mental model of how the service is being hosted.

@rneswold Do I understand correctly that you merged #79 that broke main and then pulled the broken main into this PR?

rneswold committed:
:memo: reformat comments

rneswold committed:
Merge branch 'main' into epics

beauremus reviewed (approved):
Thanks for all the work! :shipit:
